File: naddu/naddu_2/c/make.c
#include <string.h>
#include "make.h"
#include "zob.h"
#include "node.h"
#include "gen.h"
#include "move.h"

static uint8_t rights_mask[128];

void make_init(void) {
  memset(rights_mask, 0x0F, 128);
  rights_mask[0x00] = 0x0F ^ RIGHTS_Q;
  rights_mask[0x07] = 0x0F ^ RIGHTS_K;
  rights_mask[0x70] = 0x0F ^ RIGHTS_q;
  rights_mask[0x77] = 0x0F ^ RIGHTS_k;
  rights_mask[0x04] = 0x0F ^ RIGHTS_K ^ RIGHTS_Q;
  rights_mask[0x74] = 0x0F ^ RIGHTS_k ^ RIGHTS_q;
}

void make_move(uint32_t move, Pos *pos) {
  int from = MOVE_FROM(move);
  int to = MOVE_TO(move);
  int piece = pos->board[from];
  int captured = pos->board[to];
  int old_rights = pos->rights;
  int old_ep = pos->ep;

  // Update halfmove clock
  if ((piece & 0x7) == PAWN || captured || (move & MOVE_FLAG_EPCAPTURE)) {
    pos->hmc = 0;
  }
  else {
    pos->hmc++;
  }

  uint64_t *piece_zob = zob_pieces[piece];
  uint64_t h = pos->hash;

  // Remove piece from 'from' square
  h ^= piece_zob[from];

  // Remove captured piece (if any)
  if (captured) {
    h ^= zob_pieces[captured][to];
  }

  pos->board[to] = piece;
  pos->board[from] = 0;
  pos->ep = 0;

  if (move & MOVE_EXTRA_MASK) {
    if (move & MOVE_PROMO_MASK) {
      int promo_piece = pos->stm | ((move & MOVE_PROMO_MASK) >> MOVE_PROMO_SHIFT);
      pos->board[to] = promo_piece;
      h ^= zob_pieces[promo_piece][to];
    }
    else if (move & MOVE_FLAG_KING) {
      pos->kings[piece >> 3] = to;
      h ^= piece_zob[to];
    }
    else if (move & MOVE_FLAG_EPMAKE) {
      pos->ep = (pos->stm == WHITE) ? to - 16 : to + 16;
      h ^= piece_zob[to];
    }
    else if (move & MOVE_FLAG_EPCAPTURE) {
      int cap_sq = (pos->stm == WHITE) ? to - 16 : to + 16;
      int cap_piece = pos->board[cap_sq];
      h ^= zob_pieces[cap_piece][cap_sq];
      pos->board[cap_sq] = 0;
      h ^= piece_zob[to];
    }
    else if (move & MOVE_FLAG_KCASTLE) {
      pos->kings[piece >> 3] = to;
      int rook = pos->board[to + 1];
      uint64_t *rook_zob = zob_pieces[rook];
      h ^= rook_zob[to + 1] ^ rook_zob[to - 1];
      pos->board[to - 1] = rook;
      pos->board[to + 1] = 0;
      h ^= piece_zob[to];
    }
    else if (move & MOVE_FLAG_QCASTLE) {
      pos->kings[piece >> 3] = to;
      int rook = pos->board[to - 2];
      uint64_t *rook_zob = zob_pieces[rook];
      h ^= rook_zob[to - 2] ^ rook_zob[to + 1];
      pos->board[to + 1] = rook;
      pos->board[to - 2] = 0;
      h ^= piece_zob[to];
    }
  }
  else {
    h ^= piece_zob[to];
  }

  pos->rights &= rights_mask[from] & rights_mask[to];

  // Update rights hash
  if (pos->rights != old_rights) {
    h ^= zob_rights[old_rights] ^ zob_rights[pos->rights];
  }

  // Update ep hash
  if (old_ep) {
    h ^= zob_ep[old_ep];
  }
  if (pos->ep) {
    h ^= zob_ep[pos->ep];
  }

  // Toggle stm
  h ^= zob_black;

  pos->stm ^= BLACK;
  pos->hash = h;
}

void do_move(const char *uci_move) {
  Node *node = &nodes[0];
  Pos *pos = &node->pos;

  gen_moves(node);

  char buf[8];
  for (int i = 0; i < node->num_moves; i++) {
    uint32_t move = node->moves[i];
    format_move(move, buf);
    if (strcmp(buf, uci_move) == 0) {
      make_move(move, pos);
      return;
    }
  }
}

File: naddu/naddu_2/src/make.js
const RIGHTS_MASK = new Uint8Array(128);
RIGHTS_MASK.fill(0xF);
RIGHTS_MASK[0x00] = 0xF ^ RIGHTS_Q;
RIGHTS_MASK[0x07] = 0xF ^ RIGHTS_K;
RIGHTS_MASK[0x70] = 0xF ^ RIGHTS_q;
RIGHTS_MASK[0x77] = 0xF ^ RIGHTS_k;
RIGHTS_MASK[0x04] = 0xF ^ RIGHTS_K ^ RIGHTS_Q;
RIGHTS_MASK[0x74] = 0xF ^ RIGHTS_k ^ RIGHTS_q;

function makeMove(move, pos) {

  const from = (move >> 8) & 0xff;
  const to = move & 0xff;
  const piece = pos.board[from];
  const captured = pos.board[to];
  const oldRights = pos.rights;
  const oldEp = pos.ep;

  // Update halfmove clock: reset on pawn move or capture, else increment
  if ((piece & 0x7) === PAWN || captured || (move & MOVE_FLAG_EPCAPTURE)) {
    pos.hmc = 0;
  }
  else {
    pos.hmc++;
  }

  const pieceZobLo = zobPiecesLo[piece];
  const pieceZobHi = zobPiecesHi[piece];

  var lo = pos.hashLo;
  var hi = pos.hashHi;

  // remove piece from 'from' square
  lo ^= pieceZobLo[from];
  hi ^= pieceZobHi[from];

  // remove captured piece (if any)
  if (captured) {
    lo ^= zobPiecesLo[captured][to];
    hi ^= zobPiecesHi[captured][to];
  }

  pos.board[to] = piece;
  pos.board[from] = 0;
  pos.ep = 0;

  if (move & MOVE_EXTRA_MASK) {
    if (move & MOVE_PROMO_MASK) {
      const promoPiece = pos.stm | (move & MOVE_PROMO_MASK) >> MOVE_PROMO_SHIFT;
      pos.board[to] = promoPiece;
      lo ^= zobPiecesLo[promoPiece][to];
      hi ^= zobPiecesHi[promoPiece][to];
    }

    else if (move & MOVE_FLAG_KING) {
      pos.kings[piece >> 3] = to;
      lo ^= pieceZobLo[to];
      hi ^= pieceZobHi[to];
    }

    else if (move & MOVE_FLAG_EPMAKE) {
      pos.ep = pos.stm === WHITE ? to - 16 : to + 16;
      lo ^= pieceZobLo[to];
      hi ^= pieceZobHi[to];
    }

    else if (move & MOVE_FLAG_EPCAPTURE) {
      const capSq = pos.stm === WHITE ? to - 16 : to + 16;
      const capPiece = pos.board[capSq];
      lo ^= zobPiecesLo[capPiece][capSq];
      hi ^= zobPiecesHi[capPiece][capSq];
      pos.board[capSq] = 0;
      lo ^= pieceZobLo[to];
      hi ^= pieceZobHi[to];
    }

    else if (move & MOVE_FLAG_KCASTLE) {
      pos.kings[piece >> 3] = to;
      const rook = pos.board[to + 1];
      const rookZobLo = zobPiecesLo[rook];
      const rookZobHi = zobPiecesHi[rook];
      lo ^= rookZobLo[to + 1] ^ rookZobLo[to - 1];
      hi ^= rookZobHi[to + 1] ^ rookZobHi[to - 1];
      pos.board[to - 1] = rook;
      pos.board[to + 1] = 0;
      lo ^= pieceZobLo[to];
      hi ^= pieceZobHi[to];
    }

    else if (move & MOVE_FLAG_QCASTLE) {
      pos.kings[piece >> 3] = to;
      const rook = pos.board[to - 2];
      const rookZobLo = zobPiecesLo[rook];
      const rookZobHi = zobPiecesHi[rook];
      lo ^= rookZobLo[to - 2] ^ rookZobLo[to + 1];
      hi ^= rookZobHi[to - 2] ^ rookZobHi[to + 1];
      pos.board[to + 1] = rook;
      pos.board[to - 2] = 0;
      lo ^= pieceZobLo[to];
      hi ^= pieceZobHi[to];
    }
  }
  else {
    lo ^= pieceZobLo[to];
    hi ^= pieceZobHi[to];
  }

  pos.rights &= RIGHTS_MASK[from] & RIGHTS_MASK[to];

  // update rights hash
  if (pos.rights !== oldRights) {
    lo ^= zobRightsLo[oldRights] ^ zobRightsLo[pos.rights];
    hi ^= zobRightsHi[oldRights] ^ zobRightsHi[pos.rights];
  }

  // update ep hash
  if (oldEp) {
    lo ^= zobEpLo[oldEp];
    hi ^= zobEpHi[oldEp];
  }
  if (pos.ep) {
    lo ^= zobEpLo[pos.ep];
    hi ^= zobEpHi[pos.ep];
  }

  // toggle stm
  lo ^= zobBlackLo;
  hi ^= zobBlackHi;

  pos.stm ^= BLACK;
  pos.hashLo = lo >>> 0;
  pos.hashHi = hi >>> 0;

}

function doMove(uciMove) {

  const node = nodes[0];
  const pos = node.pos;

  genMoves(node);

  for (let i = 0; i < node.numMoves; i++) {
    const move = node.moves[i];
    if (formatMove(move) === uciMove) {
      makeMove(move, pos);
      return;
    }
  }
}
